hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestPositionBasedMergingFallback.scala

```diff
@@ -0,0 +1,192 @@
+import org.apache.hudi.common.model.HoodieRecordMerger
+import org.apache.hudi.common.util
+import org.apache.hudi.config.HoodieWriteConfig
+import org.apache.hudi.testutils.HoodieSparkClientTestBase
+import org.apache.hudi.util.JFunction
+import org.apache.spark.sql.SaveMode.{Append, Overwrite}
+import org.apache.spark.sql.SparkSessionExtensions
+import org.apache.spark.sql.hudi.HoodieSparkSessionExtension
+import org.apache.spark.sql.internal.SQLConf
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.{AfterEach, BeforeEach}
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.{Arguments, MethodSource}
+
+import java.util.function.Consumer
+
+class TestPositionBasedMergingFallback extends HoodieSparkClientTestBase {
+  override def getSparkSessionExtensionsInjector: util.Option[Consumer[SparkSessionExtensions]] =
+    toJavaOption(
+      Some(
+        JFunction.toJavaConsumer((receiver: SparkSessionExtensions) => new HoodieSparkSessionExtension().apply(receiver)))
+    )
+
+  @BeforeEach override def setUp(): Unit = {
+    initPath()
+    initSparkContexts()
+    sparkSession.conf.set(SQLConf.PARQUET_RECORD_FILTER_ENABLED.key, "true")
+    initTestDataGenerator()
+    initHoodieStorage()
+  }
+
+  @AfterEach override def tearDown(): Unit = {
+    cleanupSparkContexts()
+    cleanupTestDataGenerator()
+    cleanupFileSystem()
+    FileSystem.closeAll()
+    System.gc()
+  }
+
+  @ParameterizedTest
+  @MethodSource(Array("testArgs"))
+  def testPositionFallback(updateWithRecordPositions: String, deleteWithRecordPositions: String, secondUpdateWithPositions: String): Unit = {
+    val columns = Seq("ts", "key", "name", "_hoodie_is_deleted")
+    val data = Seq(
+      (10, "1", "A", false),
+      (10, "2", "B", false),
+      (10, "3", "C", false),
+      (10, "4", "D", false),
+      (10, "5", "E", false))
+
+    val inserts = sparkSession.createDataFrame(data).toDF(columns: _*)
+    inserts.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
+      option(PRECOMBINE_FIELD.key(), "ts").
+      option("hoodie.table.name", "test_table").
+      option(TABLE_TYPE.key(), "MERGE_ON_READ").
+      option(HoodieStorageConfig.LOGFILE_DATA_BLOCK_FORMAT.key(), "parquet").
+      option(DataSourceWriteOptions.RECORD_MERGER_STRATEGY.key(), HoodieRecordMerger.DEFAULT_MERGER_STRATEGY_UUID).
+      option(DataSourceWriteOptions.RECORD_MERGER_IMPLS.key(), "org.apache.hudi.HoodieSparkRecordMerger").
+      option(HoodieReaderConfig.FILE_GROUP_READER_ENABLED.key(), "true").
+      option(HoodieWriteConfig.WRITE_RECORD_POSITIONS.key(), "true").
+      mode(Overwrite).
+      save(basePath)
+
+    val updateData = Seq((11, "1", "A_1", false), (9, "2", "B_1", false))
+
+    val updates = sparkSession.createDataFrame(updateData).toDF(columns: _*)
+
+    updates.write.format("hudi").
+      option(RECORDKEY_FIELD.key(), "key").
```

Comment:
let's extract the options to separate methods such as `getWriterOptions` and `getReaderOptions` which can take the same arguments as passed to the test.